### PR TITLE
Implement GRPC client as transport

### PR DESCRIPTION
**Pull request by sorru94**

### CODE REVIEW

File: examples/grpc/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+SPDX-License-Identifier: Apache-2.0
+-->
+
+# Astarte device SDK Python GRPC example
+This is an example of how to use the device SDK to connect a to an existing Astarte message hub
+instance and handle datastream messages from/to the node.
+
+## Prerequisites
+
+An instance of the message hub should be running on an available machine.
```
Comment (rgallor) [inline examples/grpc/README.md:13]: You could put a reference to the message hub setup webpage.

**Review by rgallor — CHANGES_REQUESTED** [review]

File: astarte/device/device_grpc.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+
+class ConnectionState(Enum):
+    """
+    Possible connection states for a device.
+    """
+
+    CONNECTING = 1
+    CONNECTED = 2
+    DISCONNECTED = 3
+
```
Comment (harlem88) [inline astarte/device/device_grpc.py:70]: The ConnecctionState enum and related methods is defined also in the `device-mqtt.py`, What do you think to move in the Abstract Class? Is this possible?

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: astarte/device/device_grpc.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+        self.__msghub_stub.Detach(self.__msghub_node)
+        self.__msghub_node = Node(
+            uuid=self._node_uuid, interface_jsons=list(self.__interfaces_bins.values())
+        )
+        stream = self.__msghub_stub.Attach(self.__msghub_node)
+        self.__stream_queue.put(stream)
+
+    def connect(self) -> None:
+        """
+        Connect the device in a synchronous manner.
```
Comment (harlem88) [inline astarte/device/device_grpc.py:170]:
```suggestion
        Connect the device in synchronous mode.
```

File: .github/workflows/e2e-test.yaml [hunk 4]
```diff
@@ -95,3 +99,113 @@ jobs:
...
+          echo "E2E_REALM=test" >> $GITHUB_ENV
+          APPENGINE_TOKEN=$(astartectl utils gen-jwt appengine)
+          echo "E2E_APPENGINE_TOKEN=$APPENGINE_TOKEN" >> $GITHUB_ENV
+          echo "E2E_APPENGINE_URL=https://api.autotest.astarte-platform.org/appengine" >> $GITHUB_ENV
+          echo "E2E_PAIRING_URL=https://api.autotest.astarte-platform.org/pairing" >> $GITHUB_ENV
+      - name: Setup env variables for device 1
+        run: |
+          DEVICE_1_ID=$(astartectl utils device-id generate-random)
+          echo "E2E_DEVICE_1_ID=$DEVICE_1_ID" >> $GITHUB_ENV
+          echo "E2E_DEVICE_3_ID=$DEVICE_1_ID" >> $GITHUB_ENV
```
Comment (harlem88) [inline .github/workflows/e2e-test.yaml:153]: Why do you use two different (1,3) device_id?

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e-test.yaml (hunk 4 above)
Comment (sorru94) [reply .github/workflows/e2e-test.yaml:153]: Ah, this is quite an ugly workaround to have the same tests used for the mqtt transport work for grpc.
Right now each test has a device number hardcoded (base is 1, persistency is 2 and reconnection is 3).
The device ID is necessary also when running the tests for grpc as HTTP transactions are used to verify the device behavior.
With grpc there is a single device, so the required environment variables have all the same value.

Comment (sorru94) [reply .github/workflows/e2e-test.yaml:153]: I'll fix this by passing the device number as a command line argument.

**Review by joshuachp — APPROVED** [review]

**Review by rgallor — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]

**Review by harlem88 — COMMENTED** [review]
Add an entry about the introducing of GRPC as transport to the changelog.

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,16 +4,22 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [0.12.1] - Unreleased
+## Unreleased
```
Comment (harlem88) [inline CHANGELOG.md:7]:
```suggestion
## [Unreleased]
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — APPROVED** [review]

**Pull request merged**